### PR TITLE
8266249: javax/swing/JPopupMenu/7156657/bug7156657.java fails on macOS

### PR DESCRIPTION
**Pull request by azuev-java**
Fixed popup position taking into account its offset
Added a lot of screenshots taking to better understand failures should they happen down the line

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266249](https://bugs.openjdk.java.net/browse/JDK-8266249): javax/swing/JPopupMenu/7156657/bug7156657.java fails on macOS

### Reviewers
 * [Jayathirth D V](https://openjdk.java.net/census#jdv) (@jayathirthrao - **Reviewer**) ⚠️ Review applies to f3589c68d1fd7dc7fb4fa2aa255315828860c29d
 * [Pankaj Bansal](https://openjdk.java.net/census#pbansal) (@pankaj-bansal - **Reviewer**)
 * [Alexander Zvegintsev](https://openjdk.java.net/census#azvegint) (@azvegint - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3844/head:pull/3844` \
`$ git checkout pull/3844`

Update a local copy of the PR: \
`$ git checkout pull/3844` \
`$ git pull https://git.openjdk.java.net/jdk pull/3844/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3844`

View PR using the GUI difftool: \
`$ git pr show -t 3844`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3844.diff">https://git.openjdk.java.net/jdk/pull/3844.diff</a>

</details>

### CODE REVIEW

**Comment by pankaj-bansal** [comment]
As this is an intermittent failure (failed once in 40 iterations as mentioned in JBS), can you please run multiple iterations of the test in CI system without JTREG_RETRY_COUNT option to confirm there is no failure now?

**Comment by azuev-java** [comment]
> As this is an intermittent failure (failed once in 40 iterations as mentioned in JBS), can you please run multiple iterations of the test in CI system without JTREG_RETRY_COUNT option to confirm there is no failure now?

Added test run link to the bug.

**Review by jayathirthrao — APPROVED** [review]
LGTM.
I had doubt whether we will have write access to test path to create new files in our CI systems, but it looks like we can. Verified it by making the test fail purposefully in our CI.

**Comment by prsadhuk** [comment]
I'm not sure about this change...Maybe changing frame location from (0,0) to middle of screen though selLocationRelativeTo() may be enough to make the test more stable in CI systems.

**Comment by azuev-java** [comment]
> I'm not sure about this change...Maybe changing frame location from (0,0) at l192 to middle of screen though selLocationRelativeTo() may be enough to make the test more stable in CI systems.

The problem here is not only in location but in the screenshot taking area outside of the popup and analyzing it. That is one of the reasons why test is unreliable - it analyzes the wrong place. And moving frame to the different position will not help either - the problem is not that system menubar overlaps the popup - it does not, the problem is that we are catching part of the menubar because we do not adjust for the offset it introduces to the screen coordinates.

**Comment by prsadhuk** [comment]
> 
> 
> > I'm not sure about this change...Maybe changing frame location from (0,0) at l192 to middle of screen though selLocationRelativeTo() may be enough to make the test more stable in CI systems.
> 
> The problem here is not only in location but in the screenshot taking area outside of the popup and analyzing it. That is one of the reasons why test is unreliable - it analyzes the wrong place. And moving frame to the different position will not help either - the problem is not that system menubar overlaps the popup - it does not, the problem is that we are catching part of the menubar because we do not adjust for the offset it introduces to the screen coordinates. Here's the screenshot, note that area we are analyzing is one outlined with cyan borders. By not adjusting the location we just catching random stuff.
> dev0scr0

OK...I will still urge to bring the frame to middle to be more safe.
BTW, greenFrame and redFrame screencapture are needed only in failure case so I guess those should be called if compare fails.
Also, dispose should be called in EDT same way we are doing a little bit afterwards..

**Review by pankaj-bansal — APPROVED** [review]

**Review by azvegint — APPROVED** [review]

File: test/jdk/javax/swing/JPopupMenu/7156657/bug7156657.java [hunk 1]
```diff
@@ -103,24 +109,56 @@ public void run() {
         Rectangle popupRectangle = Util.invokeOnEDT(new Callable<Rectangle>() {
             @Override
             public Rectangle call() throws Exception {
-                return popupMenu.getBounds();
+                return new Rectangle(popupMenu.getLocationOnScreen(),
+                        popupMenu.getSize());
```
Comment (mrserb) [inline test/jdk/javax/swing/JPopupMenu/7156657/bug7156657.java:113]: Does it mean that the getBounds method returns the wrong coordinate?

**Pull request closed without merging**